File: java/src/poop_project/PAMFormatter.java
package poop_project;

import java.io.File;
import java.io.FileWriter;
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

public class PAMFormatter extends Formatter {

	public PAMFormatter() {
		
	}

	@Override
	public void write(String location, Image image) throws IOException {
		
		String header = "P7\n"+
						"WIDTH "+image.getWWidth()+"\n"+
						"HEIGHT "+image.getHHeight()+"\n"+
						"DEPTH 4\n"+
						"MAXVAL 255\n"+
						"TUPLTYPE RGB_ALPHA\n"+
						"ENDHDR\n";
		
		String data = header;
		
		StringBuilder str = new StringBuilder(data);
		
		
		for (int i = 0; i < image.getHHeight();i++ ) {
			for (int j = 0; j < image.getWWidth();j++ ) {
				Pixel pixel_write = image.calculatePixel(i, j);
				str.append((char)pixel_write.getRed()).append((char)pixel_write.getGreen()).append((char)pixel_write.getBlue()).append((char)pixel_write.getAlpha());
				//data += ((char)pixel_write.getRed()) + ((char)pixel_write.getGreen()) + ((char)pixel_write.getBlue()) + ((char)pixel_write.getAlpha());
			}
		}
		File output = new File(location);
		output.createNewFile();
		FileWriter writer = new FileWriter(output);
		writer.write(str.toString());
		writer.flush(); writer.close();
	}
	

	@Override
	public Layer read(String location) throws IOException {
		
		byte[] data = Files.readAllBytes(Paths.get(location));
		String buffer = new String(data);
		
		//std::regex rx("P7WIDTH (.*)HEIGHT (.*)DEPTH (.*)MAXVAL (.*)TUPLTYPE (.*)ENDHDR");//P7WIDTH 128HEIGHT 128DEPTH 4MAXVAL 255TUPLTYPE RGB_ALPHAENDHDR
		
		Pattern regex =Pattern.compile("P7\nWIDTH ([0-9]*)\nHEIGHT ([0-9]*)\nDEPTH ([0-9]*)\nMAXVAL ([0-9]*)\nTUPLTYPE (.*)\nENDHDR\n(.*)");
		Matcher matcher = regex.matcher(buffer);
		//System.out.print(buffer);
		
		boolean flag = matcher.matches();
		System.out.println(flag);
		
		
		int W = Integer.parseInt(matcher.group(1));
		int H = Integer.parseInt(matcher.group(2));
		
//		Layer layer = new Layer(Integer.parseInt(matcher.group(1)),Integer.parseInt(matcher.group(2)));
		Layer layer = new Layer(W,H);
		if(flag) {
			
			byte[] pixels = matcher.group(6).getBytes();
			
			
			for(int i = Integer.parseInt(matcher.group(2))-1 ; i >=0;i--) {
				for(int j =0 ; j < Integer.parseInt(matcher.group(1));j++) {
					layer.pixels.get(i).set(j,
									  new Pixel((pixels[(i*W+j)*4+0] & 255),
												(pixels[(i*W+j)*4+1] & 255),
												(pixels[(i*W+j)*4+2] & 255),
												(pixels[(i*W+j)*4+3] & 255)));
				}
			}
			
			
		}else {
			System.out.println("Greska pri citnaju PAM formata");
		}
		
		
		return layer;
	}

}

File: cpp/PAMFormatter.cpp
#include"PAMFormatter.h"
#include<fstream>
#include<regex>
#include<iostream>
PAMFormatter::PAMFormatter(const std::string& name) : Formatter(name) {}

std::ostream& operator<<(std::ostream& stream, const Pam_read& pam) {
	stream << "P7" << std::endl << "WIDTH: " << pam.WIDTH << std::endl
		<< "HEIGHT: " << pam.HEIGHT << std::endl
		<< "DEPTH: " << pam.DEPTH << std::endl
		<< "MAXVAL: " << pam.MAXVAL << std::endl
		<< "TUPLTYPE: " << pam.TUPLTYPE << std::endl
		<< "ENDHDR ";
	return stream;;
}

void PAMFormatter::read(Layer& layer) {

	std::ifstream input{ this->path,std::ios_base::binary };
	if (!input.is_open()) throw std::runtime_error("Error opening pam file.");


	std::string  line, header_str;
	std::regex rx("P7WIDTH (.*)HEIGHT (.*)DEPTH (.*)MAXVAL (.*)TUPLTYPE (.*)ENDHDR");//P7WIDTH 128HEIGHT 128DEPTH 4MAXVAL 255TUPLTYPE RGB_ALPHAENDHDR
	std::smatch result;

	for (int i = 0; i < 7; i++) {
		std::getline(input, line);
		header_str += line;
	}

	if (std::regex_match(header_str, result, rx)) {
		HR.WIDTH = stoi(result.str(1));
		HR.HEIGHT = stoi(result.str(2));
		HR.DEPTH = stoi(result.str(3));
		HR.MAXVAL = stoi(result.str(4));
		HR.TUPLTYPE = result.str(5);

	}
	else {
		throw std::runtime_error("neprepoznat header-regex");
	}

	layer = Layer(HR.WIDTH, HR.HEIGHT);

	char* pixel = new char[4]; //ovde cemo procitati nase pixele kako bi ih rasporedili

	for (int i = 0; i < HR.HEIGHT; i++) {
		for (int j = 0; j < HR.WIDTH; j++) {
			input.read(pixel, 4);
			layer[i][j] = Pixel((uint8_t)pixel[0], (uint8_t)pixel[1], (uint8_t)pixel[2], (uint8_t)pixel[3]); // RGBA
		}
	}


}




void PAMFormatter::write(const std::string& name, Image& image) {
	
	std::ofstream output{ name,std::ios_base::binary };

	if (!output.is_open()) throw std::runtime_error("Error with outputfile.");

	output.write(HW.P7, sizeof(HW.P7));				//P7
	output.write(HW.space, sizeof(HW.space));		//space
	output.write(HW.width, sizeof(HW.width));		//WIDTH
	output << image.getWidth();						//width val
	output.write(HW.space, sizeof(HW.space));		//space
	output.write(HW.height, sizeof(HW.height));		//HEIGHT
	output << image.getHeight();					//height val
	output.write(HW.space, sizeof(HW.space));		//space
	output.write(HW.depth, sizeof(HW.depth));		//DEPTH
	output << 4;									//depth val
	output.write(HW.space, sizeof(HW.space));		//space
	output.write(HW.maxval, sizeof(HW.maxval));		//MAXVAL	
	output << 255;									//maxval val
	output.write(HW.space, sizeof(HW.space));		//space
	output.write(HW.tupltype, sizeof(HW.tupltype));	//tupltype
	output.write(HW.rgb_alpha, sizeof(HW.rgb_alpha));//type
	output.write(HW.space, sizeof(HW.space));		//space
	output.write(HW.endhrd, sizeof(HW.endhrd));		//ENDHDR
	output.write(HW.space, sizeof(HW.space));		//space


	std::regex rx("[^.]*(.*)");
	std::smatch result;
	std::regex_match(this->path,result,rx );

//	if ((result.str(1) == ".bmp" && (layer.getReverse()==false))   ||  (result.str(1) == ".pam" && (layer.getReverse() == true))) layer.reverse();


	char* pixel = new char[4];

	for (unsigned i = 0; i < image.getHeight(); i++) {
		for (unsigned j = 0; j < image.getWidth(); j++) {
			Pixel pixel_write = image.calculatePixel(i, j);
			pixel[2] = pixel_write.getBlue();
			pixel[1] = pixel_write.getGreen();
			pixel[0] = pixel_write.getRed();
			pixel[3] = pixel_write.getAlpha();
			output.write(pixel, 4);
		}
	}

	

	output.close();







}